### PR TITLE
OGM-1237 Test keep duplicates and preserve order

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/OGM-1237

### CODE REVIEW

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+			session.save( meeting );
+		} );
+
+		inTransaction( session -> {
+			Meeting load = session.load( Meeting.class, NAME );
+
+			// an embedded list with a defined @OrderColumn keeps duplicates
+			assertThat( load.getPhones() ).containsExactly( PHONE, PHONE, PHONE );
+
+			// without an @OrderCoulmn it may or may not keep duplicate
```
Comment (DavideD) [inline core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: s/Coulm/Column/

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+			assertThat( load.getAddresses() ).containsOnly( ADDRESS, ADDRESS_ALT1, ADDRESS_ALT2 );
+
+			// a join table list with a defined @OrderColumn keeps duplicates
+			assertThat( load.getSpeakers() ).containsExactly( speaker, speaker, speaker );
+
+			// without an @OrderCoulmn it may or may not not keep duplicate
```
Comment (DavideD) [inline core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: s/Coulm/Column/

Comment (DavideD) [reply core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: s/duplicate/duplicates/

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+			assertThat( load.getParticipants() ).containsOnly( participant0, participant1, participant2 );
+		} );
+	}
+
+	@Test
+	public void testPreserveOrderOnAssociationItems() {
```
Comment (DavideD) [inline core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: Isn't this test the same as List#testOrderedList

Comment (DavideD) [reply core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: Let's keep it anyway, but let's be careful not too have too many similar tests

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import org.hibernate.ogm.utils.TestForIssue;
+
+import org.junit.After;
+import org.junit.Test;
+
+/**
+ * When a {@link javax.persistence.OrderColumn} is defined on association mapped as {@link List},
+ * order and duplicates will be preserved over different loads.
+ *
+ * Otherwise that will not guarantee. It will depend on the specific datastore behavior.
```
Comment (DavideD) [inline core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: Can we keep this javadoc simpler?
Something like:
{quote}
Tests for associations and collections of elements containing duplicates, with or without @OrderColumn.
{quote}

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java (hunk 3 above)
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: We verify the order here, instead of the duplicate keeping.

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java (hunk 4 above)
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: Of course we can

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java (hunk 3 above)
Comment (DavideD) [reply core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: > We verify the order here, instead of the duplicate keeping.

 The same thing we do in `List#testOrderedList`.
Also this issue was about testing duplicates and there are no duplicates here.

Just saying, this class become much shorter and simpler if we only focus on what we are working on without loosing in test coverage

Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: I think it isn't the same. In the latter we check order.

**Comment by fax4ever** [comment]
I put the issue in preview state: I'm updating the docs.

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java (hunk 3 above)
Comment (DavideD) [reply core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: Are you saying we are not checking the order of the elements here? https://github.com/hibernate/hibernate-ogm/blob/2a7c097408f6b8fcc771cd5bee570b8c588eed5d/core/src/test/java/org/hibernate/ogm/backendtck/associations/collection/types/ListTest.java#L62

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+		} );
+
+		inTransaction( session -> {
+			Meeting load = session.load( Meeting.class, NAME );
+
+			// an embedded list with a defined @OrderColumn guarantees the order
+			assertThat( load.getPhones() ).containsExactly( PHONE, PHONE_ALT1, PHONE_ALT2 );
+
+			// without an @OrderColumn it doesn't guarantee the order
+			assertThat( load.getAddresses() ).containsOnly( ADDRESS, ADDRESS_ALT1, ADDRESS_ALT2 );
```
Comment (DavideD) [inline core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: > assertThat( load.getAddresses() ).containsOnly( ADDRESS, ADDRESS_ALT1, ADDRESS_ALT2 );

I don't think we need this test.
We already plenties of tests under `org.hibernate.ogm.backendtck.embeddable`.

For this issue I would change this to:

```
meeting.addAddress( ADDRESS );
meeting.addAddress( ADDRESS );
meeting.addAddress( ADDRESS );

...
assertThat( load.getAddresses() ).as(
"Hibernate OGM does not supporty duplicates for a collection of embeddebles (issue ...). If you've solved this issue, congratulations! Please, update the jira and this test accordingly and thanks a lot"
).containsOnly( ADDRESS  );
```

File: core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java (hunk 3 above)
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/backendtck/elementcollection/DuplicatesAndOrderOfAssociationItemsTest.java]: Whether I try to substitute:

> assertThat( load.getAddresses() ).containsOnly( ADDRESS, ADDRESS_ALT1, ADDRESS_ALT2 );

with:

> assertThat( load.getAddresses() ).containsExactly( ADDRESS, ADDRESS_ALT1, ADDRESS_ALT2 );

I got this:

> org.junit.ComparisonFailure: 
Expected :['1 avenue des Champs Elysees, Paris, France', 'Piazza del Colosseo, 1, Rome, Italy', 'Charing Cross, London WC2N 5DU, UK']
Actual   :['Charing Cross, London WC2N 5DU, UK', '1 avenue des Champs Elysees, Paris, France', 'Piazza del Colosseo, 1, Rome, Italy']

File: documentation/manual/src/main/asciidoc/modules/mapping.asciidoc [hunk 6]
```diff
@@ -120,3 +120,20 @@ store 1 million tuples in the association key.
...
+Alternatively, it is possible to tell Hibernate to handle the order, using the annotations: `@javax.persistence.OrderColumn` and `@javax.persistence.OrderBy`.
+For more information on order see the Hibernate ORM guide http://docs.jboss.org/hibernate/orm/5.3/userguide/html_single/Hibernate_User_Guide.html#collections-list.
+
+==== Keep duplicates of the association items
+
+Duplicate items on association can be lost, even if it used a `List` as a collection type.
+Without the use of the `javax.persistence.OrderColumn` annotation, the responsibility to keep duplicates is left to the underlining datastore.
+Most of them are idempotent on keys, it means that duplicates will be lost in save. Others aren't, so they will be able to keep duplicates.
+
+Using the JPA annotation `@javax.persistence.OrderColumn`, Hibernate will take care of the duplicates, as well as order.
```
Comment (DavideD) [inline documentation/manual/src/main/asciidoc/modules/mapping.asciidoc]: Maybe it's just me, but the way you are wording this makes it look scarier than it actually is. If you don't mind I will try to phrase it slightly differently.

Comment (fax4ever) [reply documentation/manual/src/main/asciidoc/modules/mapping.asciidoc]: Sure I don't mind ;)

**Comment by DavideD** [comment]
I've created a new PR with a couple of addditional commits https://github.com/hibernate/hibernate-ogm/pull/1090
